File: packages/validator-go/validator/legacy/validator.go
package legacy

import (
	"sort"
	"strconv"
	"strings"
)

// Validator is the main validator struct
type Validator struct {
	spec            Spec
	rules           map[string]RuleFunc
	conditionParser *ConditionParser
}

// ruleMode controls which rules apply in a validation pass.
type ruleMode int

const (
	// ruleModeAll applies every rule (scalar fields, group-level rules).
	ruleModeAll ruleMode = iota
	// ruleModeArray applies only array-level rules to a multiple field's array.
	ruleModeArray
	// ruleModeItem applies only item-level rules to each array element.
	ruleModeItem
)

// arrayLevelRules operate on the whole array of a multiple field,
// never on its individual items.
var arrayLevelRules = map[string]bool{
	"required": true,
	"mincount": true,
	"maxcount": true,
	"unique":   true,
}

// pathReferenceRules receive raw path strings as parameters; their string
// params are never evaluated as condition expressions.
var pathReferenceRules = map[string]bool{
	"equalTo":  true,
	"notEqual": true,
	"enddate":  true,
}

// literalParamRules receive literal strings (regex patterns, accept lists) as
// parameters; their string params are never treated as condition expressions
// (ternary expressions are still evaluated). An accept param such as ".jpg"
// starts with a dot and would otherwise be misread as a relative field
// reference and the rule skipped.
var literalParamRules = map[string]bool{
	"match":   true,
	"pattern": true,
	"accept":  true,
}

// ruleEntry is a rule name/param pair in declaration order.
type ruleEntry struct {
	name  string
	param interface{}
}

// NewValidator creates a new validator instance
func NewValidator(spec Spec) *Validator {
	return &Validator{
		spec:            spec,
		rules:           DefaultRules(),
		conditionParser: NewConditionParser(),
	}
}

// Validate validates all data against the spec
func (v *Validator) Validate(data map[string]interface{}) *ValidationResult {
	result := &ValidationResult{
		IsValid: true,
		Errors:  []ValidationError{},
	}

	v.validateFields(v.spec.Fields, data, data, []string{}, result)

	return result
}

// ValidateField validates a single field
func (v *Validator) ValidateField(path string, value interface{}, allData map[string]interface{}) *string {
	pathParts := StringToPath(path)

	field := v.findFieldByPath(pathParts)
	if field == nil {
		return nil // No field definition found, skip validation
	}

	result := &ValidationResult{IsValid: true, Errors: []ValidationError{}}
	v.validateFieldRules(field, value, pathParts, allData, result, ruleModeAll)
	if len(result.Errors) > 0 {
		msg := result.Errors[0].Message
		return &msg
	}
	return nil
}

// AddRule adds a custom validation rule
func (v *Validator) AddRule(name string, fn RuleFunc) {
	v.rules[name] = fn
}

// validateFields recursively validates fields in spec declaration order.
// data: current scope data for value access
// rootData: full form data for condition evaluation
func (v *Validator) validateFields(fields []Field, data map[string]interface{}, rootData map[string]interface{}, currentPath []string, result *ValidationResult) {
	for i := range fields {
		field := &fields[i]
		fieldPath := AppendToPath(currentPath, field.Name)

		var value interface{}
		if data != nil {
			value = data[field.Name]
		}

		// Skip hidden fields (display_switch / display_target)
		if !v.shouldDisplay(field, rootData, fieldPath) {
			continue
		}

		// Group fields (have nested properties)
		if len(field.Fields) > 0 {
			v.validateGroupField(field, value, fieldPath, rootData, result)
			continue
		}

		// Multiple leaf field with array data: array-level rules first,
		// then item-level rules per element.
		if field.Multiple {
			if arr, ok := value.([]interface{}); ok {
				if !v.validateFieldRules(field, arr, fieldPath, rootData, result, ruleModeArray) {
					for idx, item := range arr {
						itemPath := AppendToPath(fieldPath, strconv.Itoa(idx))
						v.validateFieldRules(field, item, itemPath, rootData, result, ruleModeItem)
					}
				}
				continue
			}
			// Non-array data on a multiple field falls through to normal validation
		}

		v.validateFieldRules(field, value, fieldPath, rootData, result, ruleModeAll)
	}
}

// validateGroupField validates a group field (nested, repeatable, or "only").
func (v *Validator) validateGroupField(field *Field, value interface{}, fieldPath []string, rootData map[string]interface{}, result *ValidationResult) {
	switch {
	case field.MultipleOnly:
		// "only" mode: single object treated like an array for wildcards
		if m, ok := value.(map[string]interface{}); ok {
			v.validateFields(field.Fields, m, rootData, fieldPath, result)
		}
		v.validateFieldRules(field, value, fieldPath, rootData, result, ruleModeAll)

	case field.Multiple:
		if arr, ok := value.([]interface{}); ok {
			// Repeatable group (array of objects)
			for idx, item := range arr {
				itemMap, _ := item.(map[string]interface{})
				if itemMap == nil {
					itemMap = map[string]interface{}{}
				}
				itemPath := AppendToPath(fieldPath, strconv.Itoa(idx))
				v.validateFields(field.Fields, itemMap, rootData, itemPath, result)
			}
			v.validateFieldRules(field, value, fieldPath, rootData, result, ruleModeAll)
		} else if m, ok := value.(map[string]interface{}); ok {
			// Repeatable group stored as object with unique keys
			keys := make([]string, 0, len(m))
			for k := range m {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			for _, k := range keys {
				itemMap, _ := m[k].(map[string]interface{})
				if itemMap == nil {
					itemMap = map[string]interface{}{}
				}
				v.validateFields(field.Fields, itemMap, rootData, AppendToPath(fieldPath, k), result)
			}
			v.validateFieldRules(field, value, fieldPath, rootData, result, ruleModeAll)
		}
		// Mismatched data shape: skip validation

	default:
		// Single nested group: validate children even when the group data
		// is missing, so required rules inside still apply.
		m, _ := value.(map[string]interface{})
		if m == nil {
			m = map[string]interface{}{}
		}
		v.validateFields(field.Fields, m, rootData, fieldPath, result)
		v.validateFieldRules(field, value, fieldPath, rootData, result, ruleModeAll)
	}
}

// validateFieldRules applies a field's rules in declaration order,
// stopping at the first error. Reports whether an error was recorded.
func (v *Validator) validateFieldRules(field *Field, value interface{}, fieldPath []string, rootData map[string]interface{}, result *ValidationResult, mode ruleMode) bool {
	pathStr := PathToString(fieldPath)

	// For number type fields, implicitly run number validation first
	// if there's no explicit number rule (to catch invalid numbers before min/max)
	if mode != ruleModeArray && field.Type == "number" && !v.hasRule(field, "number") && !isEmpty(value) {
		_, isArr := value.([]interface{})
		_, isMap := value.(map[string]interface{})
		if !isArr && !isMap {
			if numberRule := v.rules["number"]; numberRule != nil {
				ctx := &ValidationContext{CurrentPath: fieldPath, FormData: rootData, FieldDef: field}
				if errMsg := numberRule(value, nil, rootData, ctx); errMsg != nil {
					result.IsValid = false
					result.Errors = append(result.Errors, ValidationError{
						Field:   pathStr,
						Rule:    "number",
						Message: v.getErrorMessage(field, "number", *errMsg, nil),
						Value:   value,
					})
					return true
				}
			}
		}
	}

	for _, entry := range v.orderedRules(field) {
		if mode == ruleModeArray && !arrayLevelRules[entry.name] {
			continue
		}
		if mode == ruleModeItem && arrayLevelRules[entry.name] {
			continue
		}

		errMsg, params := v.applyRule(entry.name, entry.param, value, fieldPath, rootData, field)
		if errMsg != nil {
			result.IsValid = false
			result.Errors = append(result.Errors, ValidationError{
				Field:   pathStr,
				Rule:    entry.name,
				Message: v.getErrorMessage(field, entry.name, *errMsg, params),
				Value:   value,
			})
			return true // Stop at first error for this field
		}
	}

	return false
}

// applyRule applies a single validation rule.
// String parameters are evaluated first: ternary expressions yield their
// branch value, and a false condition expression skips the rule.
func (v *Validator) applyRule(ruleName string, ruleParam interface{}, value interface{}, fieldPath []string, rootData map[string]interface{}, field *Field) (*string, []string) {
	if s, ok := ruleParam.(string); ok && !pathReferenceRules[ruleName] {
		// Regex-param rules (match/pattern) may contain "?:" sequences;
		// only space-delimited ternaries count as expressions for them.
		isTernary := IsTernaryExpression(s)
		if literalParamRules[ruleName] {
			isTernary = isTernary && strings.Contains(s, " ? ") && strings.Contains(s, " : ")
		}
		if isTernary {
			resolved := v.conditionParser.EvaluateTernaryString(s, rootData, fieldPath)
			if resolved == nil {
				return nil, nil
			}
			if b, ok := resolved.(bool); ok && !b {
				return nil, nil
			}
			ruleParam = resolved
		} else if !literalParamRules[ruleName] && IsConditionExpression(s) {
			conditionMet, err := v.conditionParser.Evaluate(s, rootData, fieldPath)
			if err != nil || !conditionMet {
				return nil, nil // Condition not met (or invalid), skip this rule
			}
			ruleParam = true
		}
	}

	// Skip if rule param is explicitly false
	if b, ok := ruleParam.(bool); ok && !b {
		return nil, nil
	}

	// For non-required fields, skip validation if value is empty.
	// mincount still applies to empty arrays (count 0).
	if ruleName != "required" && ruleName != "mincount" && isEmpty(value) {
		return nil, nil
	}

	ruleFn, ok := v.rules[ruleName]
	if !ok {
		// Check if it's a custom rule defined in the spec
		if customRule, ok := v.spec.Rules[ruleName]; ok {
			ctx := &ValidationContext{CurrentPath: fieldPath, FormData: rootData, FieldDef: field}
			return v.applyCustomRule(&customRule, value, rootData, ctx), nil
		}
		return nil, nil // Unknown rule, skip
	}

	ctx := &ValidationContext{CurrentPath: fieldPath, FormData: rootData, FieldDef: field}
	params := parseRuleParams(ruleParam)
	return ruleFn(value, params, rootData, ctx), params
}

// orderedRules returns a field's rules in declaration order.
// Without declaration order info, required comes first and the remaining
// rules follow alphabetically (deterministic fallback).
func (v *Validator) orderedRules(field *Field) []ruleEntry {
	var entries []ruleEntry
	seen := map[string]bool{}

	appendRule := func(name string) {
		if seen[name] {
			return
		}
		if param, ok := field.Rules[name]; ok {
			entries = append(entries, ruleEntry{name: name, param: param})
			seen[name] = true
		}
	}

	if field.Rules != nil {
		for _, name := range field.RuleOrder {
			appendRule(name)
		}

		if len(seen) < len(field.Rules) {
			appendRule("required")
			rest := make([]string, 0, len(field.Rules))
			for name := range field.Rules {
				if !seen[name] {
					rest = append(rest, name)
				}
			}
			sort.Strings(rest)
			for _, name := range rest {
				appendRule(name)
			}
		}
	}

	// Legacy top-level required attribute
	if field.Required != nil && !seen["required"] {
		entries = append([]ruleEntry{{name: "required", param: field.Required}}, entries...)
	}

	return entries
}

// hasRule reports whether the field declares a rule.
func (v *Validator) hasRule(field *Field, name string) bool {
	if field.Rules == nil {
		return false
	}
	_, ok := field.Rules[name]
	return ok
}

// shouldDisplay evaluates display_switch / display_target.
// display_switch semantics: condition string evaluated against form data;
// boolean false means always hidden (validation skipped), true always shown.
func (v *Validator) shouldDisplay(field *Field, rootData map[string]interface{}, fieldPath []string) bool {
	switch ds := field.DisplaySwitch.(type) {
	case nil:
		// Not set: visible
	case bool:
		if !ds {
			return false
		}
	case string:
		if ds != "" {
			visible, err := v.conditionParser.Evaluate(ds, rootData, fieldPath)
			if err != nil || !visible {
				return false
			}
		}
	default:
		if !isTruthy(field.DisplaySwitch) {
			return false
		}
	}

	if field.DisplayTarget != "" {
		target := v.resolveFieldReference(field.DisplayTarget, fieldPath, rootData)
		switch tv := target.(type) {
		case nil:
			return false
		case string:
			if tv == "" {
				return false
			}
		case bool:
			if !tv {
				return false
			}
		case []interface{}:
			if len(tv) == 0 {
				return false
			}
		case map[string]interface{}:
			if len(tv) == 0 {
				return false
			}
		}
	}

	return true
}

// resolveFieldReference resolves a field reference for display_target.
// Dot-prefixed references resolve relative to the current path; bare names
// resolve as siblings first, then from the root.
func (v *Validator) resolveFieldReference(ref string, fieldPath []string, rootData map[string]interface{}) interface{} {
	if strings.HasPrefix(ref, ".") {
		value, err := v.conditionParser.EvaluateValue(ref, rootData, fieldPath)
		if err != nil {
			return nil
		}
		return value
	}

	segments := StringToPath(ref)
	if len(fieldPath) > 0 {
		sibling := append(append([]string{}, fieldPath[:len(fieldPath)-1]...), segments...)
		if value := getNestedValue(rootData, sibling); value != nil {
			return value
		}
	}
	return getNestedValue(rootData, segments)
}

// applyCustomRule applies a custom rule from spec
func (v *Validator) applyCustomRule(rule *Rule, value interface{}, allData map[string]interface{}, ctx *ValidationContext) *string {
	// Pattern matching
	if rule.Pattern != "" {
		matchRule := v.rules["match"]
		if matchRule != nil {
			errMsg := matchRule(value, []string{rule.Pattern}, allData, ctx)
			if errMsg != nil {
				if rule.Message != "" {
					return &rule.Message
				}
				return errMsg
			}
		}
	}

	// Min value
	if rule.Min != nil {
		minRule := v.rules["min"]
		if minRule != nil {
			errMsg := minRule(value, []string{strconv.Itoa(*rule.Min)}, allData, ctx)
			if errMsg != nil {
				if rule.Message != "" {
					return &rule.Message
				}
				return errMsg
			}
		}
	}

	// Max value
	if rule.Max != nil {
		maxRule := v.rules["max"]
		if maxRule != nil {
			errMsg := maxRule(value, []string{strconv.Itoa(*rule.Max)}, allData, ctx)
			if errMsg != nil {
				if rule.Message != "" {
					return &rule.Message
				}
				return errMsg
			}
		}
	}

	return nil
}

// parseRuleParams parses parameters from a rule value.
// String values stay intact (regex patterns may contain ':' or ',').
func parseRuleParams(ruleValue interface{}) []string {
	switch val := ruleValue.(type) {
	case bool:
		return nil
	case int:
		return []string{strconv.Itoa(val)}
	case int64:
		return []string{strconv.FormatInt(val, 10)}
	case float64:
		return []string{strconv.FormatFloat(val, 'f', -1, 64)}
	case string:
		return []string{val}
	case []interface{}:
		var params []string
		for _, item := range val {
			params = append(params, toString(item))
		}
		return params
	case []string:
		return val
	default:
		return nil
	}
}

// getErrorMessage resolves the final error message for a rule failure.
// Custom field messages override defaults; {0}, {1}, ... placeholders are
// substituted with rule parameters.
func (v *Validator) getErrorMessage(field *Field, ruleName string, defaultMsg string, params []string) string {
	msg := defaultMsg
	if field.Messages != nil {
		if custom, ok := field.Messages[ruleName]; ok {
			msg = custom
		}
	}
	for i, p := range params {
		msg = strings.ReplaceAll(msg, "{"+strconv.Itoa(i)+"}", p)
	}
	return msg
}

// findFieldByPath finds a field definition by path
func (v *Validator) findFieldByPath(path []string) *Field {
	if len(path) == 0 {
		return nil
	}

	return v.findFieldInList(v.spec.Fields, path, 0)
}

// findFieldInList recursively finds a field in a list
func (v *Validator) findFieldInList(fields []Field, path []string, depth int) *Field {
	if depth >= len(path) {
		return nil
	}

	targetName := path[depth]

	// Skip numeric indices (array elements)
	if _, err := strconv.Atoi(targetName); err == nil {
		if depth+1 < len(path) {
			// Continue searching in the same fields (for repeatable groups)
			return v.findFieldInList(fields, path, depth+1)
		}
		return nil
	}

	for i := range fields {
		field := &fields[i]
		if field.Name == targetName {
			if depth == len(path)-1 {
				return field
			}
			if field.Fields != nil {
				return v.findFieldInList(field.Fields, path, depth+1)
			}
		}
	}

	return nil
}

// getNestedValue retrieves a nested value from data by path.
func getNestedValue(data map[string]interface{}, path []string) interface{} {
	if len(path) == 0 {
		return data
	}

	var current interface{} = data
	for _, segment := range path {
		if current == nil {
			return nil
		}

		switch v := current.(type) {
		case map[string]interface{}:
			current = v[segment]
		case []interface{}:
			idx, err := strconv.Atoi(segment)
			if err != nil || idx < 0 || idx >= len(v) {
				return nil
			}
			current = v[idx]
		default:
			return nil
		}
	}

	return current
}

// GetSpec returns the spec
func (v *Validator) GetSpec() Spec {
	return v.spec
}

// GetRules returns the registered rules
func (v *Validator) GetRules() map[string]RuleFunc {
	return v.rules
}

File: packages/validator-rust/src/legacy/validator.rs
//! Main validator: declaration-order traversal, first-error-per-field,
//! implicit number, multiple/array handling, display_switch/display_target.
//! Ports validator-go/validator/validator.go.

use crate::legacy::condition_parser::ConditionParser;
use crate::legacy::rules::{default_rules, RuleFn, ValidationContext};
use crate::legacy::ternary::{
    evaluate_ternary_string, is_condition_expression, is_ternary_expression,
};
use crate::legacy::types::{Field, Spec, ValidationError, ValidationResult};
use crate::legacy::value::{format_float, get_nested_value, is_empty, is_truthy, to_string};
use serde_json::Value;
use std::collections::{HashMap, HashSet};

/// ruleMode controls which rules apply in a validation pass.
#[derive(Clone, Copy, PartialEq)]
enum RuleMode {
    All,
    Array,
    Item,
}

fn array_level_rules() -> &'static HashSet<&'static str> {
    use std::sync::OnceLock;
    static S: OnceLock<HashSet<&'static str>> = OnceLock::new();
    S.get_or_init(|| {
        ["required", "mincount", "maxcount", "unique"]
            .into_iter()
            .collect()
    })
}

fn path_reference_rules() -> &'static HashSet<&'static str> {
    use std::sync::OnceLock;
    static S: OnceLock<HashSet<&'static str>> = OnceLock::new();
    S.get_or_init(|| ["equalTo", "notEqual", "enddate"].into_iter().collect())
}

fn literal_param_rules() -> &'static HashSet<&'static str> {
    use std::sync::OnceLock;
    static S: OnceLock<HashSet<&'static str>> = OnceLock::new();
    S.get_or_init(|| ["match", "pattern", "accept"].into_iter().collect())
}

struct RuleEntry {
    name: String,
    param: Value,
}

/// Validates form data against a [`Spec`], producing a [`ValidationResult`].
///
/// Fields are checked in declaration order, stopping at the first error per
/// field — identical to the JS/PHP/Go validators so the same spec and data
/// yield the same result in every language. Construct with [`Validator::new`]
/// (a parsed spec is available via [`crate::legacy::parse_spec`]).
pub struct Validator {
    spec: Spec,
    rules: HashMap<&'static str, RuleFn>,
    condition_parser: ConditionParser,
}

impl Validator {
    /// Builds a validator for `spec`, registering the built-in rule set.
    pub fn new(spec: Spec) -> Self {
        Validator {
            spec,
            rules: default_rules(),
            condition_parser: ConditionParser::new(),
        }
    }

    /// Validates the whole `data` object against the spec and returns the
    /// aggregate result (all field errors, first-error-per-field).
    pub fn validate(&mut self, data: &Value) -> ValidationResult {
        let mut result = ValidationResult::default();
        let fields = self.spec.fields.clone();
        self.validate_fields(&fields, data, data, &[], &mut result);
        result
    }

    fn validate_fields(
        &mut self,
        fields: &[Field],
        data: &Value,
        root_data: &Value,
        current_path: &[String],
        result: &mut ValidationResult,
    ) {
        for field in fields {
            let mut field_path = current_path.to_vec();
            field_path.push(field.name.clone());

            let value = data
                .as_object()
                .and_then(|m| m.get(&field.name))
                .cloned()
                .unwrap_or(Value::Null);

            if !self.should_display(field, root_data, &field_path) {
                continue;
            }

            // Group fields (have nested properties).
            if !field.fields.is_empty() {
                self.validate_group_field(field, &value, &field_path, root_data, result);
                continue;
            }

            // Multiple leaf field with array data.
            if field.multiple {
                if let Value::Array(arr) = &value {
                    let stopped = self.validate_field_rules(
                        field,
                        &value,
                        &field_path,
                        root_data,
                        result,
                        RuleMode::Array,
                    );
                    if !stopped {
                        for (idx, item) in arr.iter().enumerate() {
                            let mut item_path = field_path.clone();
                            item_path.push(idx.to_string());
                            self.validate_field_rules(
                                field,
                                item,
                                &item_path,
                                root_data,
                                result,
                                RuleMode::Item,
                            );
                        }
                    }
                    continue;
                }
                // Non-array data falls through to normal validation.
            }

            self.validate_field_rules(field, &value, &field_path, root_data, result, RuleMode::All);
        }
    }

    fn validate_group_field(
        &mut self,
        field: &Field,
        value: &Value,
        field_path: &[String],
        root_data: &Value,
        result: &mut ValidationResult,
    ) {
        let children = field.fields.clone();
        if field.multiple_only {
            if let Value::Object(m) = value {
                let mv = Value::Object(m.clone());
                self.validate_fields(&children, &mv, root_data, field_path, result);
            }
            self.validate_field_rules(field, value, field_path, root_data, result, RuleMode::All);
            return;
        }

        if field.multiple {
            match value {
                Value::Array(arr) => {
                    for (idx, item) in arr.iter().enumerate() {
                        let item_map = if item.is_object() {
                            item.clone()
                        } else {
                            Value::Object(Default::default())
                        };
                        let mut item_path = field_path.to_vec();
                        item_path.push(idx.to_string());
                        self.validate_fields(&children, &item_map, root_data, &item_path, result);
                    }
                    self.validate_field_rules(
                        field,
                        value,
                        field_path,
                        root_data,
                        result,
                        RuleMode::All,
                    );
                }
                Value::Object(m) => {
                    // Repeatable group stored as object with unique keys.
                    let mut keys: Vec<String> = m.keys().cloned().collect();
                    keys.sort();
                    for k in keys {
                        let item = m.get(&k).cloned().unwrap_or(Value::Null);
                        let item_map = if item.is_object() {
                            item
                        } else {
                            Value::Object(Default::default())
                        };
                        let mut item_path = field_path.to_vec();
                        item_path.push(k);
                        self.validate_fields(&children, &item_map, root_data, &item_path, result);
                    }
                    self.validate_field_rules(
                        field,
                        value,
                        field_path,
                        root_data,
                        result,
                        RuleMode::All,
                    );
                }
                _ => {}
            }
            return;
        }

        // Single nested group: validate children even when data is missing.
        let m = if value.is_object() {
            value.clone()
        } else {
            Value::Object(Default::default())
        };
        self.validate_fields(&children, &m, root_data, field_path, result);
        self.validate_field_rules(field, value, field_path, root_data, result, RuleMode::All);
    }

    /// Returns true if an error was recorded (caller stops further passes).
    fn validate_field_rules(
        &mut self,
        field: &Field,
        value: &Value,
        field_path: &[String],
        root_data: &Value,
        result: &mut ValidationResult,
        mode: RuleMode,
    ) -> bool {
        let path_str = field_path.join(".");

        // Implicit number validation for number-type fields without an explicit
        // number rule.
        if mode != RuleMode::Array
            && field.field_type == "number"
            && !field.has_rule("number")
            && !is_empty(value)
            && !value.is_array()
            && !value.is_object()
        {
            if let Some(number_rule) = self.rules.get("number") {
                let ctx = ValidationContext {
                    current_path: field_path,
                    form_data: root_data,
                };
                if let Some(err_msg) = number_rule(value, &[], root_data, &ctx) {
                    result.is_valid = false;
                    result.errors.push(ValidationError {
                        field: path_str.clone(),
                        rule: "number".to_string(),
                        message: self.get_error_message(field, "number", &err_msg, &[]),
                        value: value.clone(),
                    });
                    return true;
                }
            }
        }

        for entry in self.ordered_rules(field) {
            if mode == RuleMode::Array && !array_level_rules().contains(entry.name.as_str()) {
                continue;
            }
            if mode == RuleMode::Item && array_level_rules().contains(entry.name.as_str()) {
                continue;
            }

            let (err_msg, params) =
                self.apply_rule(&entry.name, &entry.param, value, field_path, root_data);
            if let Some(msg) = err_msg {
                result.is_valid = false;
                result.errors.push(ValidationError {
                    field: path_str.clone(),
                    rule: entry.name.clone(),
                    message: self.get_error_message(field, &entry.name, &msg, &params),
                    value: value.clone(),
                });
                return true;
            }
        }

        false
    }

    /// apply_rule resolves string params (ternary/condition) then runs the rule.
    /// Returns (error message, params used for message substitution).
    fn apply_rule(
        &mut self,
        rule_name: &str,
        rule_param: &Value,
        value: &Value,
        field_path: &[String],
        root_data: &Value,
    ) -> (Option<String>, Vec<String>) {
        let mut effective_param = rule_param.clone();

        if let Value::String(s) = rule_param {
            if !path_reference_rules().contains(rule_name) {
                let mut is_ternary = is_ternary_expression(s);
                if literal_param_rules().contains(rule_name) {
                    is_ternary = is_ternary && s.contains(" ? ") && s.contains(" : ");
                }
                if is_ternary {
                    match evaluate_ternary_string(
                        &mut self.condition_parser,
                        s,
                        root_data,
                        field_path,
                    ) {
                        None => return (None, vec![]),
                        Some(Value::Bool(false)) => return (None, vec![]),
                        Some(resolved) => effective_param = resolved,
                    }
                } else if !literal_param_rules().contains(rule_name) && is_condition_expression(s) {
                    match self.condition_parser.evaluate(s, root_data, field_path) {
                        Ok(true) => effective_param = Value::Bool(true),
                        _ => return (None, vec![]),
                    }
                }
            }
        }

        // Skip if rule param is explicitly false.
        if let Value::Bool(false) = effective_param {
            return (None, vec![]);
        }

        // Skip non-required rules on empty values (mincount still applies).
        if rule_name != "required" && rule_name != "mincount" && is_empty(value) {
            return (None, vec![]);
        }

        let rule_fn = match self.rules.get(rule_name) {
            Some(f) => *f,
            None => {
                // Custom rule from spec.
                if let Some(custom) = self.spec.rules.get(rule_name).cloned() {
                    let msg = self.apply_custom_rule(&custom, value, field_path, root_data);
                    return (msg, vec![]);
                }
                return (None, vec![]);
            }
        };

        let ctx = ValidationContext {
            current_path: field_path,
            form_data: root_data,
        };
        let params = parse_rule_params(&effective_param);
        (rule_fn(value, &params, root_data, &ctx), params)
    }

    fn apply_custom_rule(
        &self,
        rule: &crate::legacy::types::CustomRule,
        value: &Value,
        field_path: &[String],
        root_data: &Value,
    ) -> Option<String> {
        let ctx = ValidationContext {
            current_path: field_path,
            form_data: root_data,
        };
        if let Some(pattern) = &rule.pattern {
            if !pattern.is_empty() {
                if let Some(match_rule) = self.rules.get("match") {
                    if let Some(err) =
                        match_rule(value, std::slice::from_ref(pattern), root_data, &ctx)
                    {
                        return Some(if !rule.message.is_empty() {
                            rule.message.clone()
                        } else {
                            err
                        });
                    }
                }
            }
        }
        if let Some(min) = rule.min {
            if let Some(min_rule) = self.rules.get("min") {
                if let Some(err) = min_rule(value, &[min.to_string()], root_data, &ctx) {
                    return Some(if !rule.message.is_empty() {
                        rule.message.clone()
                    } else {
                        err
                    });
                }
            }
        }
        if let Some(max) = rule.max {
            if let Some(max_rule) = self.rules.get("max") {
                if let Some(err) = max_rule(value, &[max.to_string()], root_data, &ctx) {
                    return Some(if !rule.message.is_empty() {
                        rule.message.clone()
                    } else {
                        err
                    });
                }
            }
        }
        None
    }

    /// ordered_rules returns rules in declaration order, with the legacy
    /// top-level required prepended when present.
    fn ordered_rules(&self, field: &Field) -> Vec<RuleEntry> {
        let mut entries: Vec<RuleEntry> = Vec::new();
        let mut seen: HashSet<String> = HashSet::new();

        if !field.rules.is_empty() {
            for name in &field.rule_order {
                if seen.contains(name) {
                    continue;
                }
                if let Some(param) = field.rules.get(name) {
                    entries.push(RuleEntry {
                        name: name.clone(),
                        param: param.clone(),
                    });
                    seen.insert(name.clone());
                }
            }

            // Deterministic fallback for rules not covered by rule_order.
            if seen.len() < field.rules.len() {
                if !seen.contains("required") {
                    if let Some(param) = field.rules.get("required") {
                        entries.push(RuleEntry {
                            name: "required".to_string(),
                            param: param.clone(),
                        });
                        seen.insert("required".to_string());
                    }
                }
                let mut rest: Vec<String> = field
                    .rules
                    .keys()
                    .filter(|k| !seen.contains(*k))
                    .cloned()
                    .collect();
                rest.sort();
                for name in rest {
                    if let Some(param) = field.rules.get(&name) {
                        entries.push(RuleEntry {
                            name: name.clone(),
                            param: param.clone(),
                        });
                        seen.insert(name);
                    }
                }
            }
        }

        // Legacy top-level required.
        if let Some(req) = &field.required {
            if !seen.contains("required") {
                entries.insert(
                    0,
                    RuleEntry {
                        name: "required".to_string(),
                        param: req.clone(),
                    },
                );
            }
        }

        entries
    }

    fn should_display(&mut self, field: &Field, root_data: &Value, field_path: &[String]) -> bool {
        match &field.display_switch {
            None => {}
            Some(Value::Bool(b)) => {
                if !b {
                    return false;
                }
            }
            Some(Value::String(s)) => {
                if !s.is_empty() {
                    match self.condition_parser.evaluate(s, root_data, field_path) {
                        Ok(true) => {}
                        _ => return false,
                    }
                }
            }
            Some(other) if !is_truthy(other) => return false,
            Some(_) => {}
        }

        if !field.display_target.is_empty() {
            let target = self.resolve_field_reference(&field.display_target, field_path, root_data);
            match target {
                None | Some(Value::Null) => return false,
                Some(Value::String(s)) if s.is_empty() => return false,
                Some(Value::Bool(false)) => return false,
                Some(Value::Array(a)) if a.is_empty() => return false,
                Some(Value::Object(o)) if o.is_empty() => return false,
                _ => {}
            }
        }

        true
    }

    fn resolve_field_reference(
        &mut self,
        reference: &str,
        field_path: &[String],
        root_data: &Value,
    ) -> Option<Value> {
        if reference.starts_with('.') {
            return self
                .condition_parser
                .evaluate_value(reference, root_data, field_path)
                .ok();
        }

        let segments: Vec<String> = reference
            .split('.')
            .filter(|s| !s.is_empty())
            .map(str::to_string)
            .collect();
        if !field_path.is_empty() {
            let mut sibling = field_path[..field_path.len() - 1].to_vec();
            sibling.extend(segments.clone());
            if let Some(v) = get_nested_value(root_data, &sibling) {
                if !v.is_null() {
                    return Some(v.clone());
                }
            }
        }
        get_nested_value(root_data, &segments).cloned()
    }

    fn get_error_message(
        &self,
        field: &Field,
        rule_name: &str,
        default_msg: &str,
        params: &[String],
    ) -> String {
        let mut msg = default_msg.to_string();
        if let Some(custom) = field.messages.get(rule_name) {
            msg = custom.clone();
        }
        for (i, p) in params.iter().enumerate() {
            msg = msg.replace(&format!("{{{}}}", i), p);
        }
        msg
    }
}

/// parse_rule_params mirrors Go parseRuleParams: scalars become a single param;
/// arrays become per-item string params; bool yields no params.
fn parse_rule_params(rule_value: &Value) -> Vec<String> {
    match rule_value {
        Value::Bool(_) => vec![],
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                vec![i.to_string()]
            } else if let Some(u) = n.as_u64() {
                vec![u.to_string()]
            } else if let Some(f) = n.as_f64() {
                vec![format_float(f)]
            } else {
                vec![]
            }
        }
        Value::String(s) => vec![s.clone()],
        Value::Array(arr) => arr.iter().map(to_string).collect(),
        _ => vec![],
    }
}
